test/integration/crud/client_bulk_write.test.ts

```diff
@@ -0,0 +1,349 @@
+import { expect } from 'chai';
+import { setTimeout } from 'timers/promises';
+
+import {
+  type CommandStartedEvent,
+  type Connection,
+  type ConnectionPool,
+  type MongoClient,
+  MongoOperationTimeoutError,
+  TimeoutContext
+} from '../../mongodb';
+import {
+  clearFailPoint,
+  configureFailPoint,
+  makeMultiBatchWrite,
+  makeMultiResponseBatchModelArray
+} from '../../tools/utils';
+import { filterForCommands } from '../shared';
+
+const metadata: MongoDBMetadataUI = {
+  requires: {
+    mongodb: '>=8.0',
+    serverless: 'forbid'
+  }
+};
+
+describe('Client Bulk Write', function () {
+  let client: MongoClient;
+
+  afterEach(async function () {
+    await client?.close();
+    await clearFailPoint(this.configuration);
+  });
+
+  describe('CSOT enabled', function () {
+    describe('when timeoutMS is set on the client', function () {
+      beforeEach(async function () {
+        client = this.configuration.newClient({}, { timeoutMS: 300 });
+        await client.connect();
+        await configureFailPoint(this.configuration, {
+          configureFailPoint: 'failCommand',
+          mode: { times: 1 },
+          data: { blockConnection: true, blockTimeMS: 1000, failCommands: ['bulkWrite'] }
+        });
+      });
+
+      it('timeoutMS is used as the timeout for the bulk write', metadata, async function () {
```

Comment:
No, the expectation is that the operation takes no longer than timeoutMS.  That we throw a timeout error is just a manifestation of this in our public API.

Reply:
That manifestation (throwing) is the expected behavior right? if the expectation is that it takes no longer than timeoutMS then shouldn't there be assertions that measure the distance between a start and end time?

Reply:
No.  The expected behavior is that the operation times out within timeoutMS.  The timeout error is how we communicate this to users but if that's all we consider the "semver API" - what is stopping us from immediately throwing a timeout error before executing any commands?

I've adjusted the tests to measure the execution time of each operation and assert it is within a range of timeoutMS.

Reply:
> The timeout error is how we communicate this to users

Isn't `MongoOperationTimeoutError` still the expected error? I thought these tests checked that but now it is `MongoError`.

> what is stopping us from immediately throwing a timeout error before executing any commands?

Yea I agree, I think this morning we realized we don't really have coverage for meeting the deadline in general because the spec tests only expect error, but we should think about introducing time-taken assertions.

Reply:
I changed them to MongoError because while timeout errors are the actual error thrown, that's not what this test is asserting.

Reply:
Should we also have tests that separately cover `client.bulkWrite` throws the expected timeout error? Are there CSOT spec tests for client.bulkWrite? that would suffice

Any other error thrown within the expected deadline isn't capturing the expected behavior, how we communicate the timeout matters if one is to write code against that behavior.